src/api/artifact/controller.go

```diff
@@ -269,26 +269,73 @@ func (c *controller) getByTag(ctx context.Context, repository, tag string, optio
 }
 
 func (c *controller) Delete(ctx context.Context, id int64) error {
-	// delete all tags that attached to the artifact
-	_, tags, err := c.tagMgr.List(ctx, &q.Query{
+	return c.deleteDeeply(ctx, id, true)
+}
+
+// "isRoot" is used to specify whether the artifact is the root parent artifact
+// the error handling logic for the root parent artifact and others is different
+func (c *controller) deleteDeeply(ctx context.Context, id int64, isRoot bool) error {
```

Comment:
Why not move the logic of checking if it requires deeply in the `Delete()` method? Then deleteDeeply only deeply deleted.

Reply:
But the checking is also needed for the child artifacts

Reply:
+1